[CALCITE-6770] Preserve column names when casts are inserted in projects

mihaibudiu opened the pull request.

ILuffZhe commented on core/src/test/java/org/apache/calcite/test/JdbcAdapterTest.java:
```diff
@@ -127,15 +127,15 @@ class JdbcAdapterTest {
                 + "FROM \"foodmart\".\"store\"\n"
                 + "WHERE \"store_id\" < 10")
-        .planHasSql("SELECT CAST(\"ENAME\" AS VARCHAR(30))\n"
+        .planHasSql("SELECT CAST(\"ENAME\" AS VARCHAR(30)) AS \"ENAME\"\n"
```
What if there are duplicate `ENAME` in this case? Will the 2nd `ENAME` be `ENAME0`?

mihaibudiu replied:
This preserves the names from the original select list that is being transformed. 
If the original list had duplicates, this one should too.
But I expect that this cannot happen at this point.

mihaibudiu replied:
I mean, when the names are inserted, the program has been validated.

ILuffZhe commented on core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java:
```diff
@@ -257,6 +257,15 @@ public static void checkActualAndReferenceFiles() {
+   * Preserve column names when casts are inserted in projects</a>. */
+  @Test void testCastNames() {
+    final String sql = "SELECT * FROM (SELECT empno, 'x' AS X FROM emp) "
+        + "UNION ALL (SELECT empno, 'xx' AS X from EMP)";
```
from `emp`?

ILuffZhe reviewed (approved):
LGTM, just left some minor questions.

mihaibudiu commented:
I have added a tiny commit to fix a bug I have introduced in https://github.com/apache/calcite/pull/4116.
No tests are affected, since the calcite simplification code removes the casts in question before they reach the bug.

mihaibudiu commented:
The change is in coerceColumnType, so I was assuming your case does not apply, since foo is not a column.

rongrong commented:
It's used in selectlist, case statement, in, and source row, all in `TypeCoercionImpl`.

rongrong commented:
> This actually breaks the following SQL:
> 
> ```
> 
> SELECT
> 
>   CASE
> 
>     WHEN ID IS NOT NULL
> 
>     THEN foo
> 
>     ELSE CAST(ID AS VARCHAR)
> 
>   END AS ID
> 
> FROM testTable;
> 
> ```
> 
> because it would rewrite the THEN clause to `THEN CAST(foo AS VARCHAR) AS foo` which is no longer valid SQL. Not all places accept identifiers accept alias.

This sql will fail.

mihaibudiu commented:
I understand that, but i could not write a test that fails. Can you give more details about the configuration of your compiler and how the error manifests itself?

mihaibudiu commented:
I figured out how to write a test, one way to do it is to configure the validator withIdentifierExpansion(true).
The rest should be easy.

mihaibudiu commented:
I have filed https://issues.apache.org/jira/browse/CALCITE-7072 and I will submit a fix shortly.
Thank you for the bug report.
But JIRA is really the place to file issues, people may not monitor closed PRs.

mihaibudiu commented:
@rongrong can you validate the fix in https://github.com/apache/calcite/pull/4439?
